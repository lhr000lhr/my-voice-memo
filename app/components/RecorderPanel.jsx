import React from 'react'
import { connect } from 'react-redux'
import { StyleSheet, Button, View } from 'react-native'
import { Audio } from 'expo-av'

import { createAction } from '../utils'

function RecorderPanel({ dispatch }) {
  const [recording, setRecording] = React.useState()
  const [soundUrl, setSoundUrl] = React.useState()
  const [sound, setSound] = React.useState()
  const [pause, setPause] = React.useState(false)

  async function startRecording() {
    try {
      console.log('Requesting permissions..')
      await Audio.requestPermissionsAsync()
      await Audio.setAudioModeAsync({
        allowsRecordingIOS: true,
        playsInSilentModeIOS: true,
      })
      console.log('Starting recording..')
      const player = await Audio.Recording.createAsync(Audio.RECORDING_OPTIONS_PRESET_HIGH_QUALITY)
      setRecording(player.recording)
      console.log('Recording started')
    } catch (err) {
      console.error('Failed to start recording', err)
    }
  }

  async function stopRecording() {
    console.log('Stopping recording..')
    setRecording(undefined)
    await recording.stopAndUnloadAsync()
    const uri = recording.getURI()
    setSoundUrl(uri)

    dispatch(createAction('records/add')({ recording: { uri } }))
    console.log('Recording stopped and stored at', uri)
  }

  async function pauseRecording() {
    console.log('Pausing recording..')
    await recording.pauseAsync()
    setPause(true)
  }

  async function resumeRecording() {
    console.log('Pausing recording..')
    await recording.startAsync()
    setPause(false)
  }

  async function playSound() {
    console.log('Loading Sound')
    const player = await Audio.Sound.createAsync({ uri: soundUrl })
    setSound(player.sound)

    console.log('Playing Sound')
    await player.sound.playAsync()
  }

  React.useEffect(
    () =>
      sound
        ? () => {
            console.log('Unloading Sound')
            sound.unloadAsync()
          }
        : undefined,
    [sound]
  )

  return (
    <View style={styles.container}>
      <Button
        title={recording ? 'Stop Recording' : 'Start Recording'}
        onPress={recording ? stopRecording : startRecording}
      />
      <Button
        title={pause ? 'Pause' : 'resume'}
        onPress={pause ? resumeRecording : pauseRecording}
      />

      <Button
        title="Play Sound"
        onPress={() => {
          playSound()
        }}
      />
    </View>
  )
}

const styles = StyleSheet.create({
  container: {
    backgroundColor: 'red',
    height: 200,
    width: '100%',
  },
})

export default connect()(RecorderPanel)
