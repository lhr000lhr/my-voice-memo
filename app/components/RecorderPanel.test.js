import React from 'react'
import renderer from 'react-test-renderer'
import RecorderPanel from './RecorderPanel'

jest.mock('expo-av', () => 'Audio')
jest.mock('react-redux', () => ({
  connect: () => Component => Component,
}))

describe('RecorderPanel', () => {
  it('renders correctly with RecorderPanel', () => {
    expect(renderer.create(<RecorderPanel />)).toMatchSnapshot()
  })
})
