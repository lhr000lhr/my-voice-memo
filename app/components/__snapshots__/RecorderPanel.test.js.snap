// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecorderPanel renders correctly with RecorderPanel 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "bottom",
    ]
  }
  style={
    Object {
      "backgroundColor": "gray",
      "width": "100%",
    }
  }
>
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Start Recording
      </Text>
    </View>
  </View>
</RNCSafeAreaView>
`;
